osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -59,45 +47,44 @@ public async Task<MultiplayerRoom> JoinRoom(long roomId)
                 throw new InvalidStateException("Can't join a room when restricted.");
             }
 
-            var state = await GetLocalUserState();
-
-            if (state != null)
+            using (var userUsage = await GetOrCreateLocalUserState())
             {
-                // if the user already has a state, it means they are already in a room and can't join another without first leaving.
-                throw new InvalidStateException("Can't join a room when already in another room.");
-            }
+                if (userUsage.Item != null)
+                {
+                    // if the user already has a state, it means they are already in a room and can't join another without first leaving.
+                    throw new InvalidStateException("Can't join a room when already in another room.");
+                }
 
-            // add the user to the room.
-            var roomUser = new MultiplayerRoomUser(CurrentContextUserId);
+                // add the user to the room.
+                var roomUser = new MultiplayerRoomUser(CurrentContextUserId);
 
-            // check whether we are already aware of this match.
-            if (!TryGetRoom(roomId, out var room))
-            {
-                room = await RetrieveRoom(roomId);
+                MultiplayerRoom room;
 
-                room.Host = roomUser;
-
-                lock (active_rooms)
+                using (var roomUsage = await ACTIVE_ROOMS.GetForUse(roomId, true))
                 {
-                    Console.WriteLine($"Tracking new room {roomId}.");
-                    active_rooms.Add(roomId, room);
+                    if (roomUsage.Item == null)
+                    {
+                        var retrievedRoom = await RetrieveRoom(roomId);
+                        retrievedRoom.Host = roomUser;
+                        roomUsage.Item = retrievedRoom;
+                    }
+
+                    room = roomUsage.Item;
+
+                    room.Users.Add(roomUser);
                 }
-            }
 
-            using (room.LockForUpdate())
-            {
-                room.Users.Add(roomUser);
                 await Clients.Group(GetGroupId(roomId)).UserJoined(roomUser);
                 await Groups.AddToGroupAsync(Context.ConnectionId, GetGroupId(roomId));
 
-                await UpdateDatabaseParticipants(room);
-            }
+                userUsage.Item = new MultiplayerClientState(Context.ConnectionId, CurrentContextUserId, roomId);
 
-            await UpdateLocalUserState(new MultiplayerClientState(roomId));
-
-            await MarkRoomActive(room);
+                // read only database operations are allowed to be done outside of the lock for performance reasons.
+                await UpdateDatabaseParticipants(room);
+                await MarkRoomActive(room);
```

Comment:
*Are* these read-only? Both of these don't seem to be...

Reply:
I mean read-only to the item, not the database. Will reword the comment.

Reply:
Actually for simplicity and consistency I'm just going to move them inside the lock for now (turns out the access to `room.Users` wasn't thread-safe).